go-controller/pkg/ovn/policy.go

```diff
@@ -787,6 +796,51 @@ func (oc *Controller) handlePeerPodSelectorDelete(gp *gressPolicy, obj interface
 	}
 }
 
+// handlePeerServiceSelectorAddUpdate adds the VIP of a service that selects
+// pods that are selected by the Network Policy
+func (oc *Controller) handlePeerServiceAddUpdate(gp *gressPolicy, obj interface{}) {
+	service := obj.(*kapi.Service)
+	if err := gp.addPeerSvcVip(service); err != nil {
+		klog.Errorf(err.Error())
+	}
+}
+
+// handlePeerServiceDelete removes the VIP of a service that selects
+// pods that are selected by the Network Policy
+func (oc *Controller) handlePeerServiceDelete(gp *gressPolicy, obj interface{}) {
+	service := obj.(*kapi.Service)
+	if err := gp.deletePeerSvcVip(service); err != nil {
+		klog.Errorf(err.Error())
+	}
+}
+
+// Watch Services that are in the same Namespace as the NP
+// To acount for hairpinning traffic
+func (oc *Controller) handlePeerService(
+	policy *knet.NetworkPolicy, gp *gressPolicy, np *namespacePolicy) {
+
+	h := oc.watchFactory.AddFilteredServiceHandler(policy.Namespace,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				// Service is matched so add VIP to addressSet
+				klog.V(5).Infof("A Service matches the namespace as the policy %s", policy.Name)
+				oc.handlePeerServiceAddUpdate(gp, obj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				// If SVC that has matched pods are deleted remove VIP
+				oc.handlePeerServiceDelete(gp, obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				// If SVC Is updated make sure same pods are still matched
+				if !reflect.DeepEqual(oldObj, newObj) {
```

Comment:
you may not need to do this if you are just doing cluster ip. iirc cluster IP cant change, and thats all you care about in this patch. @danwinship or @aojea can correct me if I'm wrong about that. If cluster IP can change, then the only thing you need to look for is if that changed, not the entire service object.

Reply:
Now that we're adding all possible IP's again I'll leave it

Reply:
I think you can change it to the same as we use in other places in the code like this:

```
if		reflect.DeepEqual(new.Spec.ExternalIPs, old.Spec.ExternalIPs) &&
		reflect.DeepEqual(new.Spec.ClusterIP, old.Spec.ClusterIP) &&
		reflect.DeepEqual(new.Spec.Type, old.Spec.Type) &&
		reflect.DeepEqual(new.Status.LoadBalancer.Ingress, old.Status.LoadBalancer.Ingress) {
		klog.V(5).Infof("Skipping service update for: %s as change does not apply to any of .Spec.Ports, "+
			".Spec.ExternalIP, .Spec.ClusterIP, .Spec.Type, .Status.LoadBalancer.Ingress", new.Name)
		return
	}
```

Reply:
That will involve some typecasting, is this acceptable? 
like so -> 

```
oldSvc := oldObj.(kapi.Service) 
newSvc := newObj.(kapi.Service)
```